fud2/fud-core/src/script/plugin.rs

```diff
@@ -72,6 +86,206 @@ impl ScriptContext {
+                None => Err(RhaiSystemError::state_ref(i).with_pos(pos).into()),
+            })
+            .collect::<RhaiResult<Vec<_>>>()?;
+        let output_names = output_names
+            .into_iter()
+            .map(|i| match i.clone().try_cast::<String>() {
+                Some(state) => Ok(state),
+                None => Err(RhaiSystemError::state_ref(i).with_pos(pos).into()),
+            })
+            .collect::<RhaiResult<Vec<_>>>()?;
+        let input_tuples = input_names.into_iter().zip(inputs).collect();
+        let output_tuples = output_names.into_iter().zip(outputs).collect();
+
+        let mut cur_op = self.cur_op.borrow_mut();
+        match *cur_op {
+            None => {
+                *cur_op = Some((
+                    name.to_string(),
+                    input_tuples,
+                    output_tuples,
+                    vec![],
+                ));
+                Ok(())
+            }
+            Some((ref old_name, _, _, _)) => {
+                Err(RhaiSystemError::began_op(old_name, name)
+                    .with_pos(pos)
+                    .into())
+            }
+        }
+    }
+
+    /// Adds a shell command to the `cur_op`.Returns and error if `begin_op` has not been called
+    /// before this `end_op` and after any previous `end_op`
+    fn add_shell(&self, pos: Position, cmd: String) -> RhaiResult<()> {
+        let mut cur_op = self.cur_op.borrow_mut();
+        match *cur_op {
+            Some(ref mut op_sig) => {
+                op_sig.3.push(cmd);
+                Ok(())
+            }
+            None => Err(RhaiSystemError::no_op().with_pos(pos).into()),
+        }
+    }
+
+    /// If `#ident` is a substring of `cmd` and `("ident", "value")` is in `mapping`, returns a
+    /// string identical to `cmd` except for all instances `#ident` replaced with `value`.
+    fn substitute_shell_text(cmd: &str, mapping: &[(&str, &str)]) -> String {
+        let mut cur = cmd.to_string();
+        for (i, v) in mapping {
+            let substr = format!("#{}", i);
+            cur = cur.replace(&substr, v);
+        }
+        cur
+    }
+
+    /// Collects an op currently being built and adds it to `bld`. Returns and error if `begin_op`
+    /// has not been called before this `end_op` and after any previous `end_op`.
+    fn end_op(
+        &self,
+        pos: Position,
+        mut bld: RefMut<DriverBuilder>,
+    ) -> RhaiResult<()> {
+        let mut cur_op = self.cur_op.borrow_mut();
+        match *cur_op {
+            Some((ref name, ref input_tuples, ref output_tuples, ref cmds)) => {
+                // Create the emitter.
+                let input_tuples_c = input_tuples.clone();
+                let output_tuples_c = output_tuples.clone();
+                let cmds = cmds.clone();
+                let name_c = name.clone();
+                let f: EmitBuildClosure =
+                    Box::new(move |e, inputs, outputs| {
```

Comment:
I think we can avoid using a Rust closure here, and it would actually make things easier to follow. Namely, I think we would take `OpSig` above (possibly renamed `FancyOp` or something better than this), or a more "self-contained" version that we construct here, and make it a struct that implements the `EmitBuild` trait. This would make it explicit exactly what data is required to do this build-code-emission.